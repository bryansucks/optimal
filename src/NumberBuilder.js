/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import Builder from './Builder';

function isNumber(value: *): boolean {
  return (typeof value === 'number');
}

export default class NumberBuilder extends Builder<number> {
  constructor(defaultValue: number = 0) {
    super('number', defaultValue);
  }

  between(min: number, max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(
        (isNumber(min) && isNumber(max)),
        'number.between() requires a minimum and maximum number.',
      );
    }

    return this.addCheck(this.checkBetween, min, max, inclusive);
  }

  checkBetween(path: string, value: *, min: number, max: number, inclusive: boolean = false) {
    if (__DEV__) {
      this.invariant(
        (
          isNumber(value) &&
          (inclusive ? (value >= min && value <= max) : (value > min && value < max))
        ),
        `Number must be between ${min} and ${max}${inclusive ? ' inclusive' : ''}.`,
        path,
      );
    }
  }

  oneOf(list: string[] = []): this {
    if (__DEV__) {
      this.invariant(
        (Array.isArray(list) && list.length > 0 && list.every(isNumber)),
        'number.oneOf() requires a non-empty array of numbers.',
      );
    }

    return this.addCheck(this.checkOneOf, list);
  }

  checkOneOf(path: string, value: *, list: string[] = []) {
    if (__DEV__) {
      this.invariant(
        (list.indexOf(value) >= 0),
        `Number must be one of: ${list.join(', ')}`,
        path,
      );
    }
  }
}

export function number(defaultValue: number = 0): NumberBuilder {
  return new NumberBuilder(defaultValue);
}
