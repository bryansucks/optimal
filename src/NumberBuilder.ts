/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Builder from './Builder';

function isNumber(value: any): boolean {
  return typeof value === 'number';
}

export default class NumberBuilder extends Builder<number | null> {
  constructor(defaultValue: number | null = 0) {
    super('number', defaultValue);
  }

  between(min: number, max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(
        isNumber(min) && isNumber(max),
        'Between requires a minimum and maximum number.',
      );
    }

    return this.addCheck(this.checkBetween, min, max, inclusive);
  }

  checkBetween(path: string, value: any, min: number, max: number, inclusive: boolean = false) {
    if (__DEV__) {
      this.invariant(
        isNumber(value) && (inclusive ? value >= min && value <= max : value > min && value < max),
        `Number must be between ${min} and ${max}${inclusive ? ' inclusive' : ''}.`,
        path,
      );
    }
  }

  gt(min: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(min), 'Greater-than requires a minimum number.');
    }

    return this.addCheck(this.checkGreaterThan, min, inclusive);
  }

  gte(min: number): this {
    return this.gt(min, true);
  }

  checkGreaterThan(path: string, value: any, min: number, inclusive: boolean = false) {
    if (__DEV__) {
      if (inclusive) {
        this.invariant(
          isNumber(value) && value >= min,
          `Number must be greater than or equal to ${min}.`,
          path,
        );
      } else {
        this.invariant(isNumber(value) && value > min, `Number must be greater than ${min}.`, path);
      }
    }
  }

  lt(max: number, inclusive: boolean = false): this {
    if (__DEV__) {
      this.invariant(isNumber(max), 'Less-than requires a maximum number.');
    }

    return this.addCheck(this.checkLessThan, max, inclusive);
  }

  lte(max: number): this {
    return this.lt(max, true);
  }

  checkLessThan(path: string, value: any, max: number, inclusive: boolean = false) {
    if (__DEV__) {
      if (inclusive) {
        this.invariant(
          isNumber(value) && value <= max,
          `Number must be less than or equal to ${max}.`,
          path,
        );
      } else {
        this.invariant(isNumber(value) && value < max, `Number must be less than ${max}.`, path);
      }
    }
  }

  oneOf(list: number[]): this {
    if (__DEV__) {
      this.invariant(
        Array.isArray(list) && list.length > 0 && list.every(item => isNumber(item)),
        'One of requires a non-empty array of numbers.',
      );
    }

    return this.addCheck(this.checkOneOf, list);
  }

  checkOneOf(path: string, value: any, list: number[]) {
    if (__DEV__) {
      this.invariant(list.indexOf(value) >= 0, `Number must be one of: ${list.join(', ')}`, path);
    }
  }
}

export function number(defaultValue: number | null = 0): NumberBuilder {
  return new NumberBuilder(defaultValue);
}
