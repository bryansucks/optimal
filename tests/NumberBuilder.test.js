import NumberBuilder, { number } from '../src/NumberBuilder';

describe('NumberBuilder', () => {
  let builder;

  beforeEach(() => {
    builder = new NumberBuilder(123);
  });

  describe('runChecks()', () => {
    it('errors if a non-number value is used', () => {
      expect(() => {
        builder.runChecks('key', 'foo');
      }).toThrowError('Invalid option "key". Must be a number.');
    });

    it('doesnt support null', () => {
      expect(() => {
        builder.runChecks('key', null);
      }).toThrowError('Invalid option "key". Must be a number.');
    });
  });

  describe('between()', () => {
    it('errors if min is a not a number', () => {
      expect(() => {
        builder.between('foo', 5);
      }).toThrowError('number.between() requires a minimum and maximum number.');
    });

    it('errors if max is a not a number', () => {
      expect(() => {
        builder.between(0, 'foo');
      }).toThrowError('number.between() requires a minimum and maximum number.');
    });

    it('adds a checker', () => {
      builder.between(0, 5);

      expect(builder.checks[1]).toEqual({
        func: builder.checkBetween,
        args: [0, 5, false],
      });
    });
  });

  describe('checkBetween()', () => {
    it('errors if value is not a number', () => {
      expect(() => {
        builder.checkBetween('key', 'foo', 0, 5);
      }).toThrowError('Invalid option "key". Number must be between 0 and 5.');
    });

    it('errors if out of range', () => {
      expect(() => {
        builder.checkBetween('key', 10, 0, 5);
      }).toThrowError('Invalid option "key". Number must be between 0 and 5.');
    });

    it('errors if out of range inclusive', () => {
      expect(() => {
        builder.checkBetween('key', 10, 0, 5, true);
      }).toThrowError('Invalid option "key". Number must be between 0 and 5 inclusive.');
    });

    it('doesnt error if in range', () => {
      expect(() => {
        builder.checkBetween('key', 3, 0, 5);
      }).not.toThrowError('Invalid option "key". Number must be between 0 and 5.');
    });

    it('doesnt error if in range inclusive', () => {
      expect(() => {
        builder.checkBetween('key', 5, 0, 5, true);
      }).not.toThrowError('Invalid option "key". Number must be between 0 and 5 inclusive.');
    });
  });

  describe('oneOf()', () => {
    it('errors if not an array', () => {
      expect(() => {
        builder.oneOf(123);
      }).toThrowError('number.oneOf() requires a non-empty array of numbers.');
    });

    it('errors if array is empty', () => {
      expect(() => {
        builder.oneOf([]);
      }).toThrowError('number.oneOf() requires a non-empty array of numbers.');
    });

    it('errors if array contains a non-number', () => {
      expect(() => {
        builder.oneOf(['foo', 123]);
      }).toThrowError('number.oneOf() requires a non-empty array of numbers.');
    });

    it('adds a checker', () => {
      builder.oneOf([123, 456, 789]);

      expect(builder.checks[1]).toEqual({
        func: builder.checkOneOf,
        args: [[123, 456, 789]],
      });
    });
  });

  describe('checkOneOf()', () => {
    it('errors if value is not in the list', () => {
      expect(() => {
        builder.checkOneOf('key', 666, [123, 456, 789]);
      }).toThrowError('Invalid option "key". Number must be one of: 123, 456, 789');
    });

    it('doesnt error if value contains token', () => {
      expect(() => {
        builder.checkOneOf('key', 123, [123, 456, 789]);
      }).not.toThrowError('Invalid option "key". Number must be one of: 123, 456, 789');
    });
  });
});

describe('number()', () => {
  it('returns a builder', () => {
    expect(number('foo')).toBeInstanceOf(NumberBuilder);
  });
});
