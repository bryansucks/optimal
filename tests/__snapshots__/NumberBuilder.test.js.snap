// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberBuilder between() errors if max is a not a number 1`] = `"Between requires a minimum and maximum number."`;

exports[`NumberBuilder between() errors if min is a not a number 1`] = `"Between requires a minimum and maximum number."`;

exports[`NumberBuilder checkBetween() errors if out of range 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5."`;

exports[`NumberBuilder checkBetween() errors if out of range inclusive 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5 inclusive."`;

exports[`NumberBuilder checkBetween() errors if value is not a number 1`] = `"Invalid field \\"key\\". Number must be between 0 and 5."`;

exports[`NumberBuilder checkGreaterThan() errors if below minimum 1`] = `"Invalid field \\"key\\". Number must be greater than 5."`;

exports[`NumberBuilder checkGreaterThan() errors if below minimum inclusive 1`] = `"Invalid field \\"key\\". Number must be greater than or equal to 5."`;

exports[`NumberBuilder checkGreaterThan() errors if value is not a number 1`] = `"Invalid field \\"key\\". Number must be greater than 5."`;

exports[`NumberBuilder checkLessThan() errors if above maximum 1`] = `"Invalid field \\"key\\". Number must be less than 5."`;

exports[`NumberBuilder checkLessThan() errors if above maximum inclusive 1`] = `"Invalid field \\"key\\". Number must be less than or equal to 5."`;

exports[`NumberBuilder checkLessThan() errors if value is not a number 1`] = `"Invalid field \\"key\\". Number must be less than 5."`;

exports[`NumberBuilder checkOneOf() errors if value is not in the list 1`] = `"Invalid field \\"key\\". Number must be one of: 123, 456, 789"`;

exports[`NumberBuilder gt() errors if min is a not a number 1`] = `"Greater-than requires a minimum number."`;

exports[`NumberBuilder gte() errors if min is a not a number 1`] = `"Greater-than requires a minimum number."`;

exports[`NumberBuilder lt() errors if max is a not a number 1`] = `"Less-than requires a maximum number."`;

exports[`NumberBuilder lte() errors if max is a not a number 1`] = `"Less-than requires a maximum number."`;

exports[`NumberBuilder oneOf() errors if array contains a non-number 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if array is empty 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder oneOf() errors if not an array 1`] = `"One of requires a non-empty array of numbers."`;

exports[`NumberBuilder runChecks() errors if a non-number value is used 1`] = `"Invalid field \\"key\\". Must be a number."`;
